state/models.go

```diff
@@ -734,3 +737,179 @@ func (w *Record) Equal(other Record) bool {
 func (w *Record) String() string {
 	return fmt.Sprintf("%s-%s", w.EventSource, w.EventName)
 }
+
+const TemplatePayloadKey = "template_payload"
+
+type TemplatePayload map[string]interface{}
+
+type TemplateExecutionRequest struct {
+	TemplatePayload TemplatePayload `json:"template_payload"`
+	ExecutionRequestCommon
+}
+
+func (t TemplateExecutionRequest) GetExecutionRequestCommon() *ExecutionRequestCommon {
+	return &t.ExecutionRequestCommon
+}
+
+func (t TemplateExecutionRequest) GetExecutionRequestCustom() *map[string]interface{} {
+	return &map[string]interface{}{
+		TemplatePayloadKey: t.TemplatePayload,
+	}
+}
+
+type Template struct {
+	TemplateID      string `json:"template_id"`
+	Type            string `json:"type"`
+	Version         int64  `json:"version"`
+	Schema          string `json:"schema"`
+	CommandTemplate string `json:"command_template"`
+	ExecutableResources
+}
+
+func (t Template) GetExecutableID() *string {
+	return &t.TemplateID
+}
+
+func (t Template) GetExecutableType() *ExecutableType {
+	et := ExecutableTypeTemplate
+	return &et
+}
+
+func (t Template) GetExecutableResources() *ExecutableResources {
+	return &t.ExecutableResources
+}
+
+func (t Template) GetExecutableCommand(req ExecutionRequest) (string, error) {
+	var (
```

Comment:
note: this is where jsonschema validation and go template strings come into play - deserves a closer look